pipeline-model-definition/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/AbstractModelDefTest.java

```diff
@@ -485,7 +485,35 @@ public WorkflowRun go() throws Exception {
             } else {
                 run = run.getParent().scheduleBuild2(0).waitForStart();
             }
-            j.assertBuildStatus(result, j.waitForCompletion(run));
+            try {
+                j.assertBuildStatus(result, j.waitForCompletion(run));
+            } catch (AssertionError e) {
+                // It appears that in some cases where the build result is set explictly, WorkflowRun.isBuilding()
+                // returns false too soon. Perhaps the reason is that we end up catching the Run in the middle of shutdown,
+                // so CpsFlowExecution.isComplete (and thus WorkflowRun.isInProgress() and WorkflowRun.isBuilding()) is false,
+                // but the WorkflowRun has not yet had its result set to whatever the result of the FlowEndNode is, so
+                // we see the manually-set result for a short amount of time.
+                // Here is an extract of logs where this happened while running BuildConditionResponderTest.postFailureAfterUnstable(),
+                // note that WorkflowRun#finish was called _after_ the build result was checked for the first time!
+                /*
+                  14.530 [test0 #1] [Pipeline] echo
+                  14.531 [test0 #1] I FAILED
+                  14.531 [test0 #1] [Pipeline] }
+                Build result did not initially match FAILURE
+                  14.583 [test0 #1] [Pipeline] // stage
+                  14.583 [test0 #1] [Pipeline] End of Pipeline
+                  14.589 [id=52]    INFO    o.j.p.workflow.job.WorkflowRun#finish: test0 #1 completed: FAILURE
```

Comment:
I think I was on the right track with https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/313#issuecomment-467040926, but was thinking about the wrong half of the method. When we have a manually-set build result, `WorkflowRun.isBuilding()` calls `WorkflowRun.isInProgress()`, and I think that method could return `false` before the final build result has been set in some cases (a somewhat subtle race condition).

I'm not sure if this would be an issue in real-world use cases, since I think would only affect someone polling `WorkflowRun.isBuilding()` in a tight loop to wait for a build to complete, but it would be nice to fix this somehow in workflow-job.